solr/test-framework/src/java/org/apache/solr/util/EmbeddedSolrServerTestRule.java

```diff
@@ -0,0 +1,273 @@
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.lucene.tests.util.LuceneTestCase;
+import org.apache.solr.SolrTestCaseJ4;
+import org.apache.solr.client.solrj.SolrServerException;
+import org.apache.solr.client.solrj.embedded.EmbeddedSolrServer;
+import org.apache.solr.client.solrj.embedded.EmbeddedSolrServer.RequestWriterSupplier;
+import org.apache.solr.client.solrj.impl.HttpClientUtil;
+import org.apache.solr.core.CloudConfig;
+import org.apache.solr.core.CoreContainer;
+import org.apache.solr.core.CoreDescriptor;
+import org.apache.solr.core.MetricsConfig;
+import org.apache.solr.core.NodeConfig;
+import org.apache.solr.core.PluginInfo;
+import org.apache.solr.core.SolrConfig;
+import org.apache.solr.metrics.reporters.SolrJmxReporter;
+import org.apache.solr.schema.IndexSchemaFactory;
+import org.apache.solr.update.UpdateShardHandlerConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class EmbeddedSolrServerTestRule extends SolrClientTestRule {
+
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  private EmbeddedSolrServer client = null;
+
+  public EmbeddedSolrServerTestRule() {}
+
+  public class Builder {
+    private Path solrHome;
+    private String schemaFile;
+    private String configFile = "solrconfig.xml";
+    private String collectionName = "collection1";
+    private RequestWriterSupplier requestWriterSupplier;
+
+    public Builder setSolrHome(Path solrHome) {
+      this.solrHome = solrHome;
+      return this;
+    }
+
+    public Builder setSchemaFile(String schemaFile) {
+      this.schemaFile = schemaFile;
+      return this;
+    }
+
+    public Builder setConfigFile(String configFile) {
+      this.configFile = configFile;
+      return this;
+    }
+
+    public Builder setCollectionName(String collectionName) {
+      this.collectionName = collectionName;
+      return this;
+    }
+
+    public Builder setRequestWriterSupplier(RequestWriterSupplier requestWriterSupplier) {
+      this.requestWriterSupplier = requestWriterSupplier;
+      return this;
+    }
+
+    public Path getSolrHome() {
+      return solrHome;
+    }
+
+    public String getSchemaFile() {
+      return schemaFile;
+    }
+
+    public String getConfigFile() {
+      return configFile;
+    }
+
+    public String getCollectionName() {
+      return collectionName;
+    }
+
+    public RequestWriterSupplier getRequestWriterSupplier() {
+      return requestWriterSupplier;
+    }
+
+    public void init() {
+
+      EmbeddedSolrServerTestRule.this.init(this);
+    }
+  }
+
+  private void init(Builder b) {
+    Path solrHome = b.getSolrHome();
+    String schemaFile = b.getSchemaFile();
+    String configFile = b.getConfigFile();
+    String collectionName = b.getCollectionName();
+
+    if (b.getRequestWriterSupplier() != null) {
```

Comment:
instead of if-else 3 branches, can't we always build basically one way?  Using the logic in this first branch.